### PR TITLE
Improve #1859, add binary searching methods

### PR DESCRIPTION
**Pull request by MakeNowJust**
See #1859.

This implementation supports to search in Int range and also Float
or some types range.

It uses integer and floating point value hack. This hack comes from [Ruby's implementation](https://github.com/ruby/ruby/blob/72ff61f4a8ae7a8bf05b0bd6f91b3b290645627c/range.c#L595-L662).

### CODE REVIEW

**Comment by MakeNowJust** [comment]
An interesting point:

```crystal
struct Time
  def +(other : Int)
    Time.new(ticks + other)
  end
end

p (Time.now..Time.new(2020, 1, 1)).bsearch{ |time| time.ticks > 500000008000000000_i64 }
# => 2015-11-06 02:36:28 +0000
```

By the way, why is not `Time#+(other : Int)` defined despite that `Time#-(other : Int)` is defined?

**Comment by asterite** [comment]
Thank you!

I don't have much time now to take a look at this, but:
* I don't understand how binary search can work for `Time` if there's no `Time#/`
* What's a real use case for doing a binary search in a range of floats? 
* `Time#-(other : Int)` seems like a mistake, it should only accept a `Time::Span` as an argument

**Comment by asterite** [comment]
I also feel there are too many macros, shouldn't be necessary. I also don't understand why there's a need for int_to_float and viceversa.

**Comment by MakeNowJust** [comment]
I think that doing binary search in a range of floats has more use cases than doing in a range of  intagers, because it can solve many mathematical problem.

```crystal
# solve x^4 = 800
(0...Float64::INFINITY).bsearch{|x| x ** 4 >= 800} # => 5.3183
```

Macros are too many, I think so too. I will try to reduce some macros.

**Comment by asterite** [comment]
I see. But in how many programs do you need to solve equations? And if you do need to do that, aren't there specialized, super optimized libraries for doing that?

I just don't want to add things to the standard library that aren't going to be used much, because then we'll have to keep them forever for backwards compatibility.

**Comment by MakeNowJust** [comment]
I reduced a macro `bsearch_search` and added some comments.

I often solve an equation on programming contest or machine learning.  Particularly on programming contest, `Range#bsearch` is very useful. (ping @cielavenir, did you think so?)

You can look discussion about `Range#bsearch` on Ruby at [here](https://bugs.ruby-lang.org/issues/4766) and [here](https://bugs.ruby-lang.org/issues/7724).  There are very interesting for me and you.

**Comment by asterite** [comment]
@MakeNowJust Thank, I now like it and see it can be useful :-)

I'll try to review this with @waj during this week, it looks good to me.

**Comment by cielavenir** [comment]
@MakeNowJust Well, perhaps your implementation solves Float conversion more cleanly.

**Comment by MakeNowJust** [comment]
@asterite I did `git rebase master`. If you are free, please review this.

File: spec/std/range_spec.cr [hunk 1]
```diff
@@ -26,6 +26,14 @@ struct RangeSpecIntWrapper
   end
 end
 
+private def assert_in_delta(expect, actual, delta)
```
Comment (asterite) [inline spec/std/range_spec.cr]: We have `be close(expected_value, delate)`. See https://github.com/crystal-lang/crystal/blob/master/src/spec/expectations.cr#L231

File: src/array.cr [hunk 2]
```diff
@@ -577,6 +577,44 @@ class Array(T)
...
   end
 
+  # By using binary search, returns the first element
+  # for which the passed block returns `true`.
+  #
+  # If the block returns `false`, the finding element exists
+  # behind. If the block returns `true`, the finding element
+  # is itself or exists infront.
+  #
+  # Binary search need sorted array, so self have to be sorted.
```
Comment (asterite) [inline src/array.cr]: Binary search need**s** a sorted array, so self **has** to be sorted

File: src/array.cr [hunk 3]
```diff
@@ -577,6 +577,44 @@ class Array(T)
...
+  #
+  # ```
+  # [2, 5, 7, 10].bsearch{ |x| x >= 4 } # => 5
+  # [2, 5, 7, 10].bsearch{ |x| x > 10 } # => nil
+  # ```
+  #
+  # Returns `nil` if the block didn't return `true` for any element.
```
Comment (asterite) [inline src/array.cr]: I would put this comment before the example. After the example it's less easy to spot

File: src/array.cr [hunk 4]
```diff
@@ -577,6 +577,44 @@ class Array(T)
...
+  end
+
+  # By using binary search, returns the index of the first element
+  # for which the passed block returns `true`.
+  #
+  # If the block returns `false`, the finding element exists
+  # behind. If the block returns `true`, the finding element
+  # is itself or exists infront.
+  #
+  # Binary search need sorted array, so self have to be sorted.
```
Comment (asterite) [inline src/array.cr]: needs, has (same as above)

File: src/array.cr [hunk 5]
```diff
@@ -577,6 +577,44 @@ class Array(T)
...
+  #
+  # ```
+  # [2, 5, 7, 10].bsearch_index{ |x, i| x >= 4 } # => 1
+  # [2, 5, 7, 10].bsearch_index{ |x, i| x > 10 } # => nil
+  # ```
+  #
+  # Returns `nil` if the block didn't return `true` for any element.
```
Comment (asterite) [inline src/array.cr]: Same as above, I'd move it before the example

**Comment by asterite** [comment]
@MakeNowJust Looks good! I think this addition will be very useful. I made you a couple of comments, mostly grammar mistakes (I make them all the time too :-( ).

Then please run `crystal tool format` so the code is properly formatted (that's why travis failed)

I'm not sure about the name `bsearch`. In other languages, mostly notable Java and C#, they use `binary_search`. I know `bsearch` comes from C, as many Ruby method names, but not everyone codes in C and I prefer explicit, readable names. Then the other one would be named `binary_search_index`, though maybe it's too long. I actually don't dislike the name `bsearch`, I just want to discuss this and see other's opinions ( @jhass, @ysbaddaden, @waj ).

After those decisions, I'll merge :-)

(And then we can discuss about adding the other behaviour of Array#bsearch that requires the block return an integer... when is that useful?)

**Comment by MakeNowJust** [comment]
@asterite Thanks. I refactored these codes.

> (And then we can discuss about adding the other behaviour of Array#bsearch that requires the block return an integer... when is that useful?)

I think it is useful if it returns `Array` or `Range`. For example:

```crystal
# Look up names starting with "jo". (`names` is sorted list.)
p names.bsearch { |name| name[0...2] <=> "jo" }
```

**Comment by asterite** [comment]
@jhass @ysbaddaden @waj What do you think, should we keep `bsearch` or use `binary_search` (and `binary_search_index`)?

**Comment by jhass** [comment]
I'm on mixed terms, `binary_search` seems okay and better than `bsearch`, but `binary_search_index` feels too long/verbose and `bsearch_index` feels somehow better. Given it's a known term for Rubyists at least and my feelings about the latter, I think I vote for `bsearch`.

**Comment by asterite** [comment]
@MakeNowJust Big thank you for this! This is a very powerful and needed feature.

A question about the bsearch where a block returns an integer. For this:

```crystal
# Look up names starting with "jo". (`names` is sorted list.)
p names.bsearch { |name| name[0...2] <=> "jo" }
```

Is this equivalent? Or, why is the above needed?

```crystal
# Look up names starting with "jo". (`names` is sorted list.)
p names.bsearch { |name| name.starts_with?("jo") }
```

**Comment by chocolateboy** [comment]
There's also `bisect` as used by git (and Python).

**Pull request merged**